doc/0003-wizarding-api.md

```diff
@@ -0,0 +1,227 @@
+         │                │               │              ┃
+         │                │   initiates   │              ┃
+         │                │    wizard     │              ┃            ┌─┐
+         │                │───with the ──▶│              ┃            │ │
+         │                │     given    ┌┴┐             ┃            │ │
+         │                │    element   │ │             ┃            │ │    Internal process
+         │                │              │ │wizarding    ┃            │ │
+         │                │              │ │process      ┃            │ │
+         │                │              │ │             ┃            └─┘
+         │                │              │ │             ┃
+         │                │              └┬┘             ┃
+         │                │               ├──┐           ┃
+         │                │               │  │           ┃
+         │                │               closes         ┃
+         │                │               itself         ┃
+         │                │               │  │           ┃
+         │                │               ◀──┘           ┃
+         │                │               │              ┃
+         X                X               X              ┃
+```
+
+##### Registration
+
+Registering a wizard plugin happens the same way as registering another plugin.
+We will add the wizard type to our `PluginSet`.
+```diff
+- export type PluginSet = { menu: Plugin[]; editor: Plugin[] };
++ export type PluginSet = { menu: Plugin[]; editor: Plugin[], wizard: Plugin[] };
+```
+
+Wizards must have two public static functions called `CanInspect` and `CanCreate`.
+`Core` will call these functions when it tries to figure out which wizard supports
+the given element and action.
+```ts
+type CanInspect: (tagName: string) => boolean
+type CanCreate: (tagName: string) => boolean
+```
+
+##### Requests
+
+We request a wizard through events.
+
+For viewing and/or editing purposes we can request an inspection.  
+The event is called `oscd-wizard-inspection-request` and its details look like this:
+```ts
+type WizardInspectionRequest = WizardRequest & {
+  element: Element,
+}
+```
+
+The other possible request is to create a new element under a parent element.  
+The event is called `oscd-wizard-creation-request` and its details look like this
+```ts
+type WizardCreationRequest = WizardRequest & {
+  parent: Element,
+  tagName?: string,
+}
+```
+
+Both event detail can contain an additional `stuff` for flexibility. `stuff` can be something
+that the wizard defines for additional configurations (e.g.: display it in readonly mode),
+or the editor can provide additional information about the element.
+
+> **⚠️ WARNING:** Please only use `stuff` if there is no other way.
+> It creates a tight coupling with a specific wizard or editor and it cannot be
+> guaranteed that your expectation about the `stuff` will be met!
```

Comment:
Why do we need to introduce an extra background hatch to the event that comes with a warning not to use it? What use case are we solving by this? "for flexibility" sounds scarily vague to me. The current problems with the original wizard API came about exactly because we designed it "for flexibility".

The proposal in the Wizard API issue from last year's April was exactly to make the event a pure declaration of intent to edit or create a particular element. For what purpose do we need something other than that?

Reply:
I would generally suggest that whether something should be editable or not should be up to the distribution. So for example, my distribution might include a wizard library that gets information about the current stage of the engineering process from the file and information about the user role from, say, an OpenID connect provider, and then shows read-only wizards for those elements which can not be edited by a user with this role at this stage of the engineering process. So if we want e.g. read only views in general, let's include a read only wizard library. This would mesh well with the declaration of user intent "I want to edit/create this element" that was the idea behind the API.

Do we believe "I want to view this element but I would be really angry if I could also edit it, so please don't allow me to do that!" to be a separate possible user intent? I currently don't see the use case for that, but if we do, I think it would be prudent to introduce a third type of wizard request just for viewing. What do you think about all this?

Reply:
I think you are right. This is more like a UX problem maybe and we have to decide on a case-by-case basis.